cli/mclass_info.c

```diff
@@ -0,0 +1,249 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2021 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <errno.h>
+
+#include <bsd/string.h>
+#include <cjson/cJSON.h>
+#include <yaml.h>
+
+#include <hse/hse.h>
+#include <hse_util/invariant.h>
+#include <hse_util/rest_client.h>
+
+#include <pidfile/pidfile.h>
+
+#define PATH_KEY            "path"
+#define USED_BYTES_KEY      "used_bytes"
+#define ALLOCATED_BYTES_KEY "allocated_bytes"
+
+int
+hse_mclass_info(const char *const kvdb_home, const char *const media_class)
+{
+    hse_err_t               err;
+    struct hse_kvdb *       kvdb = NULL;
+    struct hse_mclass_info  info;
+    yaml_emitter_t          emitter = {};
+    yaml_event_t            event;
+    int                     rc = 0;
+    struct pidfile          content;
+    char                    url[128];
+    char                    buf[1024];
+    cJSON *                 root;
+    HSE_MAYBE_UNUSED size_t n;
+
+    INVARIANT(kvdb_home);
+    INVARIANT(media_class);
+
+    err = hse_kvdb_open(kvdb_home, 0, NULL, &kvdb);
+    if (!err) {
+        err = hse_kvdb_mclass_info_get(kvdb, media_class, &info);
+    } else if (err && hse_err_to_errno(err) == EBUSY) {
+        rc = pidfile_deserialize(kvdb_home, &content);
+        if (rc)
+            goto out;
+
+        rc = snprintf(url, sizeof(url), "kvdb/%s/mclass/%s/info", content.alias, media_class);
+        if (rc < 0) {
+            rc = EBADMSG;
+            goto out;
+        }
+
+        assert(rc < sizeof(url));
+
+        err = curl_get(url, content.socket.path, buf, sizeof(buf));
+        if (err)
+            goto out;
+
+        root = cJSON_Parse(buf);
+        if (!root) {
+            if (cJSON_GetErrorPtr()) {
+                rc = EINVAL;
+                fprintf(stderr, "Failed to parse the response for media class info: %s", cJSON_GetErrorPtr());
```

Comment:
I'm hitting this error path when getting mclass info on an open KVDB.  This message is printed on stderr, as is the output of log_errx() from the failed call to hse_kvdb_open on line 40.  The output on my terminal is a bit garbled:
```
[HSE] hse_kvdb_open: Failed to open KVDB pidfile (/var/hse/fast0/xx/kvdb.pid): lib/binding/kvdb_interface.c:376: Device or resource busy
Failed to parse the response for media class info:
```
I presume the response is an empty string.

Reply:
You ended up running into a bug that I have just fixed locally. I am not really sure what the best course of action is regarding the log statement there.